### PR TITLE
Closes #91 length attribute from max data length 

### LINKED ISSUE #91: Allotted length: maximum length of the variable
The length for the submitted data should be set to the maximum length of the variable across all datasets per FDA Study Data Technical Conformance Guide.

In the CDISC webinar ‘Define-XML Office Hours’, they also mentioned that the length in metadata (planned length) might be different than the one in dataset (actual length) 

image

How xportr_length() could take this into account?  See below discsussion

## Definition of Done
- [ ] xportr_length() Provides the option to select either metadata length or data length.  If the length is set to 'data,' each variable's length will be the maximum length of the data, regardless of the value in the metadata.
- [ ] Include a message in xport_length() to alert the user when the data length is shorter than the length specified in the metadata. This will allow the user to update the metadata length if they wish to reduce the file size.

### PR DESCRIPTION
**Pull request by cpiraux**
### Thank you for your Pull Request!

We have developed a Pull Request template to aid you and our reviewers. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the xportr codebase remains robust and consistent.

### The scope of `{xportr}`

`{xportr}`'s scope is to enable R users to write out submission compliant `xpt` files that can be delivered to a Health Authority or to downstream validation software programs. We see labels, lengths, types, ordering and formats from a dataset specification object (SDTM and ADaM) as being our primary focus. We also see messaging and warnings to users around applying information from the specification file as a primary focus. Please make sure your Pull Request meets this **scope of {xportr}**. If your Pull Request moves beyond this scope, please get in touch with the `{xportr}` team on [slack](https://pharmaverse.slack.com/archives/C030EB2M4GM) or create an issue to discuss.

Please check off each task box as an acknowledgment that you completed the task. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

### Changes Description

- Add a function argument to get the length either from metadata or data.
- Add a utility function to get the maximum data length per variable, facilitating the automation of metadata length updates.

### Task List

- [x] The spirit of xportr is met in your Pull Request
- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Summary of changes filled out in the above Changes Description. Can be removed or left blank if changes are minor/self-explanatory.
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Use `styler` package and functions to style files accordingly.
- [x] Updated relevant unit tests or have written new unit tests. See our [Wiki](https://github.com/atorus-research/xportr/wiki/Style-Guide-for-Unit-Tests) for conventions used in this package.
- [x] Creation/updated relevant roxygen headers and examples. See our [Wiki](https://github.com/atorus-research/xportr/wiki/Style-Guide-for-Roxygen-Headers) for conventions used in this package.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Run `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new/updated functions occur on the "Reference" page.
- [x] Update NEWS.md if the changes pertain to a user-facing function (i.e. it has an @export tag) or documentation aimed at users (rather than developers)
- [x] Make sure that the pacakge version in the NEWS.md and DESCRIPTION file is same. Don't worry about updating the version because it will be auto-updated using the `vbump.yaml` CI.
- [x] Address any updates needed for vignettes and/or templates
- [x] Link the issue Development Panel so that it closes after successful merging.
- [x] Fix merge conflicts
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by bms63** [comment]
Hi @cpiraux going to have to resolve merge conflicts - a big PR was just merged.  Let me know if you need help

File: R/length.R [hunk 1]
```diff
@@ -66,6 +72,7 @@
 xportr_length <- function(.df,
                           metadata = NULL,
                           domain = NULL,
+                          length = "metadata",
                           verbose = getOption("xportr.length_verbose", "none"),
                           metacore = deprecated()) {
   if (!missing(metacore)) {
```
Comment (averissimo) [inline R/length.R:78]: Suggestion to use `match.args` just like other R functions (check `?stats::cor`)

The first element becomes the default value

```suggestion
                          length = "metadata",
                          verbose = getOption("xportr.length_verbose", "none"),
                          metacore = deprecated()) {
  length <- match.args(length)
  if (!missing(metacore)) {
```

Documentation will look like this:

image

File: R/messages.R [hunk 2]
```diff
@@ -181,3 +181,31 @@ var_ord_msg <- function(reordered_vars, moved_vars, verbose) {
...
+
+#' Utility for data Lengths
+#'
+#' @param max_length Dataframe with data and metadata length
+#' @param verbose Provides additional messaging for user
+#'
+#' @return Output to Console
+
+max_length_msg <- function(max_length, verbose) {
+  if (nrow(max_length) > 0) {
```
Comment (averissimo) [inline R/messages.R:193]: Add assertion on parameters

```suggestion
  assert_data_frame(max_length)
  assert_choice(verbose, choices = .internal_verbose_choices)
  
  if (nrow(max_length) > 0) {
```

**Review by averissimo — COMMENTED** [review]
Some comments on parameter assertion

**Comment by bms63** [comment]
@cpiraux I was missing around with this PR.  I noticed that one line had `match.args` and not `match.arg` which was causing the R-CMD checks to fail.

Now this new one is showing up.  I ran out of time to investigate.  Hoping you see the underlying issue!! 

image

**Comment by bms63** [comment]
@averissimo could you take a look at how we are using `match.arg` I am not able to get it to work.

@cpiraux I had to put in your argument into the vignettes and update the example for `xportr_length()` and things are working again!!  yay!

Some thoughts:

1) Does `xportr_lenght(lenght = "metadata")` need to be integrated into `xportr_metadata` and `xportr_write_all()` functions? @EeethB  something for us to ponder (lets address in separate issue/PR if we think it needs to be updated.)

2) @cpiraux probably shoudl build an example in the documents (function and vignettes) that allows the user to specify `length = "data"` and show user how it works

3) I think the `length` argument should be re-visited as I found it a bit confusing.  Maybe `variable_source_length` or `source_length`, `length_source`?

Okay!!  this one is almost to the finish line.  Way to go!!

**Comment by averissimo** [comment]
:facepalm: I had a typo on the suggestion, sorry about the `match.args` -> `match.args`

@bms63 Traced the issue to the "new" order of parameters. vignettes, have variations of the snippet below where they assume that the 3rd element is the `verbose` instead of `metadata`.

```R
adsl %>%
  xportr_type(var_spec, "ADSL", "message") %>%
  xportr_length(var_spec, "ADSL", "message")
#...
```

File: R/length.R [hunk 3]
```diff
@@ -62,12 +68,14 @@
 #'   length = c(10, 8)
 #' )
 #'
-#' adsl <- xportr_length(adsl, metadata, domain = "adsl")
+#' adsl <- xportr_length(adsl, metadata, domain = "adsl", length = "metadata")
 xportr_length <- function(.df,
                           metadata = NULL,
                           domain = NULL,
+                          length = c("metadata", "data"),
```
Comment (averissimo) [inline R/length.R:75]: Should we add this parameter after `verbose`?

I understand and support the reason for the current order, however, an error was triggered on our vignettes due to parameter order and I'm a bit concerned about our users facing the same problem

```r
# example
xportr_length(adsl, metadata, "adsl", "message")
#> Error in match.arg(length) : 'arg' should be one of “metadata”, “data”
```

Thoughts?

File: NEWS.md [hunk 4]
```diff
@@ -21,6 +21,8 @@
 
 * All core functions can be run together by using new function `xportr()` (#137)
 
+*New argument in `xportr_length()` allows selection between the length from metadata, as previously done, or from the calculated maximum length per variable when `length_source` is set to “data” (#91)
```
Comment (bms63) [inline NEWS.md:24]:
```suggestion
* New argument in `xportr_length()` allows selection between the length from metadata, as previously done, or from the calculated maximum length per variable when `length_source` is set to “data” (#91)
```

File: R/length.R [hunk 5]
```diff
@@ -56,12 +68,14 @@
 #'   length = c(10, 8)
 #' )
 #'
-#' adsl <- xportr_length(adsl, metadata, domain = "adsl")
+#' adsl <- xportr_length(adsl, metadata, domain = "adsl", length = "metadata")
```
Comment (bms63) [inline R/length.R:71]:
```suggestion
#' adsl <- xportr_length(adsl, metadata, domain = "adsl", length_source = "metadata")
```

File: tests/testthat/test-length.R
Comment (bms63) [inline tests/testthat/test-length.R:1]: All these will need to have the argument updated to `length_source`

File: vignettes/deepdive.Rmd [hunk 6]
```diff
@@ -194,7 +194,7 @@ To help reduce these repetitive calls, we have created `xportr_metadata()`. A us
 adsl %>%
   xportr_metadata(var_spec, "ADSL") %>%
   xportr_type() %>%
-  xportr_length() %>%
+  xportr_length(length = "metadata") %>%
```
Comment (bms63) [inline vignettes/deepdive.Rmd:197]: wondering if this can be removed since metadata is default??

**Review by bms63 — CHANGES_REQUESTED** [review]
Very close!!  Just need to update where ever length is used to length_source

**Review by bms63 — APPROVED** [review]
@cpiraux You are amazing!!  Thanks for seeing this one through for us.

**Pull request merged**